### ISSUE TITLE
cache citations using (only) CSL JSON when using the pandoc filter alone

### ISSUE THREAD

**Issue by slippycheeze** [manubot/manubot#258]
G'day.  I'm using the manubot pandoc citation filter without the rest of the tooling, as it best fits the target audience for the document – a business setting.  Overall it works wonderfully, and I'm very pleased with being able to cite resources including arbitrary webpages with minimal hassle.

## summary of the request
1. `pandoc-manubot-filter` should store all downloaded citations in a CSL JSON file.
2. `pandoc-manubot-filter` should read that same CSL JSON if it exists, but no error if it doesn't.
3. `pandoc-manubot-filter` should add new automatic citations that are not already present in the CSL JSON file.
4. `pandoc-manubot-filter` should not update citations that are present in the CSL JSON file.
     - ...or, at least, it should not update the "accessed on" date of the items, other updates might be OK.

Having this behind a flag would be fine.  This is the default behaviour of `pandoc-url2cite`, and works well (for me) – manubot has a much better range of citation identifiers, and is more robust, so I'd like to keep using it, if possible, than fall back to `pandoc-url2cite`.

Also, if you think this is a reasonable request, but don't want to implement it, I'm happy to submit an implementation.  I'd rather know it was acceptable in principal first, though.

## details of this request

When citing arbitrary webpages the "accessed on" date is important, because it is the only persistent identifier for which version of the content was used.  For accuracy is is also useful to be able to manually set that date, if I actually used a previous version of the page, or whatever.

The "accessed on" date for arbitrary webpage citations (eg: `[@https://example.org/]`) is generated when the data is fetched, a reasonable choice – but one that means I need to use the same citation data each time, rather than having manubot fetch it each time the document is built.

The `manubot-requests-cache-path` option lets me store that data in an sqlite database of binary blobs, and then reuses the same data each time.  This solves the problem of getting a stable date for when a website was accessed, but it comes with a few problems along the way:

1. the citation data is not human-readable or editable in any reasonable way.
2. wiping the cache means *all* the records are fetched again, updating the "accessed on" date for *all* webpages, and
3. the URL is hashed, making it very difficult to delete a single cached record, even if I wanted to do so.

I can also write a CSL JSON file containing the resolved references, using `manubot-output-bibliography`, which is human readable and editable, and is easy enough to remove a single citation from.  manubot doesn't read that data by default, though, so it isn't a solution unless manubot is also told to read it with the `bibliography` metadata setting, but:

1. manubot adds "manubot read this from the file ..." to the `notes` field every time, so that field grows larger with every run.  (this is the biggest problem)
2. this emits a warning the first time through, because the CSL JSON file give in `bibliography` doesn't exist yet.

**Comment by dhimmel** [comment]
Thanks @slippycheeze for the report and amazing job learning about all of the `pandoc-manubot-cite` filter features and options.

Based on the available options:

https://github.com/manubot/manubot/blob/5d9e8e4277b1e7e2a72de81e3b3796e76a504163/manubot/pandoc/cite_filter.py#L155-L160

I think your idea to combine the following options is a great one:

```markdown
---
bibliography: path/to/bibliography.json
manubot-output-bibliography: path/to/bibliography.json
...
```

> manubot adds "manubot read this from the file ..." to the notes field every time

Ah yes that is annoying. Here is the guilty code:

https://github.com/manubot/manubot/blob/217e51473f1fd1c6427803676b3c70d44314bb93/manubot/process/bibliography.py#L60-L64

Solutions that come to mind:

1. don't add manubot notes to CSL JSON if those notes already exist.
2. add an option to disable manubot from adding to notes at all (although adding the id information is important for using it as `--bibliography` input).

Let me look more into 1.

**Comment by dhimmel** [comment]
In https://github.com/manubot/manubot/pull/260, I propose a solution which avoids adding duplicate lines of text to a CSL Item note. I didn't want a fancier note-based solution because the CSL JSON schema will be adding a custom field soon which we could use a dictionary / object to encode manubot generated information in a more structured and machine-readable format.

> this emits a warning the first time through, because the CSL JSON file give in bibliography doesn't exist yet

This is still an issue. @slippycheeze how big of a problem is this for you?

One idea would be to create a new metadata field like `manubot-bibliography` that would both be read for input (if exists) and used for output.

**Comment by slippycheeze** [comment]
> This is still an issue. @slippycheeze how big of a problem is this for you?

Probably 3 out of 5, I guess?  I have `manubot-fail-on-errors: true` in my default configuration, because I prefer a hard failure to a corrupted document when I didn't notice the error.  That means that I get a failed run the first time around, though the filter does write the JSON data with all the resolved references.  So it's "only" a cosmetic issue, and won't prevent the code working.

Beyond here is opinion, based on my  assumption that most people would benefit from having the local CSL-JSON file with the auto-magically resolved citation data.  Given that, two things:

The `bibliography` command line argument will completely replace the value from the metadata file, or the pandoc defaults file.  That means it's really easy to accidentally break the manubot caching by adding my own bibliography data extras, unless I remember that fact.

Two, I can't think of any reason the cache would hurt people, I'd probably add `manubot-cache-file` which would default to something reasonable ("`manubot-bibliography.csl.json`"?), and allow it to be either a string (alternate file) or `false` to just turn off the cache.  So most people get the most reasonable default behaviour ("use a human editable cache"), and folks can still change it if they do have a need.

Demo of the problem described 
```
] pandoc -d html --csl=apa test.md
[INFO] Running filter pandoc-manubot-cite
## ERROR
load_bibliography: error reading 'citations.csl.json'.
[Errno 2] No such file or directory: 'citations.csl.json'

## CRITICAL
Failure: exiting with code 1 due to logged errors
Error running filter pandoc-manubot-cite:
Filter returned error status 1
```

I'd attach my defaults, but apparently I'm forbidden, so the relevant portions:
```yaml
filters:
  - pandoc-manubot-cite
  - citeproc
metadata:
  bibliography: citations.csl.json
  manubot-output-bibliography: citations.csl.json
  manubot-fail-on-errors: true
```

**Comment by dhimmel** [comment]
> I'd probably add `manubot-cache-file` which would default to something reasonable

I think I'm leaning towards `manubot-bibliography-cache` or just `manubot-bibliography`. Since manubot does things other than citations / bibliographies, I'd like option names that will remain unambiguous.

Here's how I'm thinking we can implement this:

1. If specified or set by default, `manubot-bibliography-cache` is added to the list of files in `metadata.bibliography` (unless it already exists in that list or does not exist on the filesystem, in which case the `metadata.bibliography` list is unmodified.
2. The final CSL Data is written back to `manubot-bibliography-cache` unless disabled.

## Questions

- Should the CSL Data cache retain CSL Items that are not cited? Not currently implemented, but if the goal is maximal caching, this might be helpful.
- Maybe yaml would be a better default format (easier to edit by hand?) than JSON (with either supported)
- Should Manubot write the CSL Data to two files if `manubot-output-bibliography` and `manubot-bibliography-cache` are both specified?

> I can't think of any reason the cache would hurt people

We initially went with a requests cache because we were frequently updating the manubot code that converts from request to CSL JSON. By caching request responses, we were able to persist the cache across manubot versions. https://github.com/manubot/manubot/commit/009184667f3aa8ac2b362aa7e3991ae4951cdde6 is a recent commit that would be able to update the output CSL JSON with the requests cache but not the CSL JSON cache. This honestly is not a huge deal and the ability to delete individual items from the CSL JSON cache in a text editor is workaround.

The other issue is the creation of an unexpected output file, which can clutter a user's system. Especially if they're running these commands from a variety of directories.

On the other hand, intermittent failure of requests to upstream bibliographic metadata resources is a nagging issue. It also can slow down builds. Manubot Rootstock enables the requests-cache by default. Having a cache enabled by default for users of `pandoc-manubot-cite` alone is probably the right call. The ability to track the cache file with git is also nice, so it can be shared across systems and users.

**Comment by dhimmel** [comment]
Opened https://github.com/manubot/manubot/pull/261. I'm still undecided about whether to set `manubot-bibliography-cache` to a default path. I think it might be better to keep it off by default until we gather a bit more feedback.

**Comment by dhimmel** [comment]
Merged #261 with support for metadata option `manubot-bibliography-cache` and released [v0.4.2](https://github.com/manubot/manubot/releases/tag/v0.4.2). @slippycheeze let us know whether this is sufficient for your needs.

One possible annoyance that comes to mind is that Manubot falls back on multiple methods for generating reference metadata for URLs. So if our zotero translation-server fails (as it sometimes does intermittently), you might end up caching sub-optimal data. But you can always delete that entry. Open to any suggestions on a good solution here.

Possibly let's make new issues for specific suggestions?

**Comment by slippycheeze** [comment]
thanks, yes, that's good.  works well for me.   

one of the reasons for asking for a CLS-JSON cache rather than sqlite+pickle was that it is (somewhat) human editable, and there are tools to convert to, eg, bibtex or other human-oriented formats.  so if I get data that isn't what I want, or has errors, I can manually fix it.   `manubot cite` is very useful for that, actually.
